curve.go

```diff
@@ -229,48 +235,29 @@ func (sc StarkCurve) MimicEcMultAir(mout, x1, y1, x2, y2 *big.Int) (x *big.Int,
 // Multiplies by m a point on the elliptic curve with equation y^2 = x^3 + alpha*x + beta mod p.
 // Assumes affine form (x, y) is spread (x1 *big.Int, y1 *big.Int) and that 0 < m < order(point).
 //
-// (ref: https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/crypto/starkware/crypto/signature/math_utils.py)
-func (sc StarkCurve) EcMult(m, x1, y1 *big.Int) (x, y *big.Int) {
-	var _ecMult func(m, x1, y1 *big.Int) (x, y *big.Int)
-
-	_add := func(x1, y1, x2, y2 *big.Int) (x, y *big.Int) {
-		yDelta := new(big.Int).Sub(y1, y2)
-		xDelta := new(big.Int).Sub(x1, x2)
-
-		m := DivMod(yDelta, xDelta, sc.P)
-
-		xm := new(big.Int).Mul(m, m)
-
-		x = new(big.Int).Sub(xm, x1)
-		x = x.Sub(x, x2)
-		x = x.Mod(x, sc.P)
-
-		y = new(big.Int).Sub(x1, x)
-		y = y.Mul(m, y)
-		y = y.Sub(y, y1)
-		y = y.Mod(y, sc.P)
+// (ref: https://www.researchgate.net/figure/Double-and-add-always-algorithm-resistant-against-SPA_fig1_48412708)
+func (sc StarkCurve) ecMult_AddAlwaysDouble(m, x1, y1 *big.Int) (x, y *big.Int) {
+	var _ecMult = func(m, x0, y0 *big.Int) (x, y *big.Int) {
+		x1, y1 := big.NewInt(0), big.NewInt(0)
+		var x2, y2 *big.Int
+		for i := 0; i <= sc.N.BitLen()-1; i++ {
+			x2, y2 = sc.Add(x0, y0, x1, y1)
+			x0, y0 = sc.Double(x0, y0)
```

Comment:
Isn't doubling suppose to happen on `Q[0]` not `P`?

Reply:
Yeah this seems to be "add then double" rather than "double then add"

Reply:
There has been a mismatch between the referenced link and the implementation, so this introduced the confusion.

In any case, I've reimplemented the algorithm to be more clear and to follow the linked literature.